src/sql/SqlService.ts

```diff
@@ -0,0 +1,290 @@
+            tryGetEnum(SqlExpectedResultType, sqlStatementOptions.expectedResultType);
+
+        if (sqlStatementOptions.hasOwnProperty('returnRawResult'))
+            tryGetBoolean(sqlStatementOptions.returnRawResult);
+    }
+
+    /**
+     * Converts datetime related wrapper classes to string to be able to serialize them. (No default serializers yet)
+     * @internal
+     * @param value
+     */
+    convertToStringIfDatetimeValue(value: any) {
+        if (value instanceof HzLocalTime || value instanceof HzLocalDate || value instanceof HzLocalDateTime) {
+            return value.toString();
+        } else if (value instanceof HzOffsetDateTime) {
+            return value.toISOString();
+        } else {
+            return value;
+        }
+    }
+
+    execute(sql: SqlStatement): SqlResult;
+    execute(sql: string, params?: any[], options?: SqlStatementOptions): SqlResult;
+    execute(sql: string | SqlStatement, params?: any[], options?: SqlStatementOptions): SqlResult {
+        let sqlStatement: SqlStatement;
+
+        if (typeof sql === 'string') {
+            sqlStatement = {
+                sql: sql
+            };
+            if (params !== undefined) { // todo: document this
+                sqlStatement.params = params;
+            }
+            if (options !== undefined) {
+                sqlStatement.options = options;
+            }
+        } else if (typeof sql === 'object') { // assume SqlStatement is passed
+            sqlStatement = sql;
+        } else {
+            throw new IllegalArgumentError('Sql can be an object or string');
+        }
+
+        try {
+            SqlServiceImpl.validateSqlStatement(sqlStatement);
+        } catch (error) {
+            throw new IllegalArgumentError(`Invalid argument given to execute(): ${error.message}`, error)
+        }
+
+        const connection = this.connectionRegistry.getRandomConnection(true);
+        if (connection === null) {
+            throw new HazelcastSqlException(
+                this.connectionManager.getClientUuid(),
+                SqlErrorCode.CONNECTION_PROBLEM,
+                'Client is not currently connected to the cluster.'
+            );
+        }
+
+        const queryId = SqlQueryId.fromMemberId(connection.getRemoteUuid());
+
+        const expectedResultType: SqlExpectedResultType =
+            sqlStatement.options?.expectedResultType ? SqlExpectedResultType[sqlStatement.options.expectedResultType]
+                : SqlServiceImpl.DEFAULT_EXPECTED_RESULT_TYPE;
+
+        try {
+            const serializedParams = [];
+            if (Array.isArray(sqlStatement.params)) { // params can be undefined
+                for (let param of sqlStatement.params) {
+                    param = this.convertToStringIfDatetimeValue(param);
+                    serializedParams.push(this.serializationService.toData(param));
+                }
+            }
+            const cursorBufferSize = sqlStatement.options?.cursorBufferSize ?
```

Comment:
same comment about the 0 value of `cursorBufferSize`

Reply:
it cannot have 0, it is validated. same for expectedResultType, validated.